### DISCUSSION TITLE
Comparing fluents between states

### DISCUSSION THREAD

**Discussion by e-cal** [AI-Planning/macq#47] [General]
Each `Step` in a trace holds its `State`, which has its list of `Fluent` objects. Each `Fluent` object stores its value as a property.

This would mean the generators handle fluent values in one of two ways:
1. Each state has its own unique `Fluent` objects
   - Determining delta state would involve cross comparing all the fluent names and object lists to figure out which ones are actually the same
   - Seems very redundant
2. Each consecutive `State` overwrites the value from state to state 
    - Only way to figure out delta state would be comparing what fluents are present and not present between 2 states (any fluents in both `State`'s lists would have the same value regardless of actual state) 

My idea would be instead to store a list of `(Fluent, value)` tuples on State and remove the `value` attribute from Fluent, or to store separate lists `fluents` and `values` that are indexed the same. 

These changes would likely break how generation currently works. Any other ideas or workarounds?

**Comment by haz** [comment]
Hrmz...I think a `State` should have a mapping from `Fluent` to it's value. Then, the diff could be something akin to...

```python
added = []
deleted = []
for f in state1:
    if state1[f] and not state2[f]:
        deleted.append(f)
    if not state1[f] and state2[f]:
        added.append(f)
return (added, deleted)
```

Means that `Fluent` needs a `__hash__` function that's tied to its name. Something like..

```python
def __hash__(self):
    return hash(str(self))
```
